prometheus/promhttp/instrument_server_test.go

```diff
@@ -0,0 +1,175 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Copyright (c) 2013, The Prometheus Authors
+// All rights reserved.
+//
+// Use of this source code is governed by a BSD-style license that can be found
+// in the LICENSE file.
+
+package promhttp
+
+import (
+	"log"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+func TestMiddlewareAPI(t *testing.T) {
+	inFlightGauge := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "inFlight",
+		Help: "Gauge.",
+	})
+
+	counter := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "test_counter",
+			Help: "Counter.",
+		},
+		[]string{"code", "method"},
+	)
+
+	histVec := prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    "latency",
+			Help:    "Histogram.",
+			Buckets: prometheus.DefBuckets,
+		},
+		[]string{"code"},
+	)
+
+	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Write([]byte("OK"))
+	})
+
+	prometheus.MustRegister(inFlightGauge, counter, histVec)
+
+	chain := InstrumentHandlerInFlight(inFlightGauge,
+		InstrumentHandlerCounter(counter,
+			InstrumentHandlerDuration(histVec, handler),
+		),
+	)
+
+	r, _ := http.NewRequest("GET", "www.example.com", nil)
+	w := httptest.NewRecorder()
+	chain.ServeHTTP(w, r)
+}
+
+func ExampleMiddleware() {
+	inFlightGauge := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "inFlight",
+		Help: "Gauge.",
+	})
+
+	counter := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "test_counter",
+			Help: "Counter.",
+		},
+		[]string{"code", "method"},
+	)
+
+	histVec := prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:    "latency",
+			Help:    "Histogram.",
+			Buckets: prometheus.DefBuckets,
+		},
+		[]string{"code"},
+	)
+
+	handler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.Write([]byte("OK"))
+	})
+
+	prometheus.MustRegister(inFlightGauge, counter, histVec)
+
+	chain := InstrumentHandlerInFlight(inFlightGauge,
+		InstrumentHandlerCounter(counter,
+			InstrumentHandlerDuration(histVec, handler),
+		),
+	)
+
+	http.Handle("/metrics", Handler())
+	http.Handle("/", chain)
+
+	if err := http.ListenAndServe(":3000", nil); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func ExampleHistogramByEndpoint() {
+	inFlightGauge := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "inFlight",
+		Help: "Gauge.",
+	})
+
+	counter := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "test_counter",
+			Help: "Counter.",
+		},
+		[]string{"code", "method"},
+	)
+
+	pushVec := prometheus.NewHistogramVec(
+		prometheus.HistogramOpts{
+			Name:        "latency",
+			Help:        "Histogram.",
+			Buckets:     []float64{.25, .5, 1, 2.5, 5, 10},
+			ConstLabels: prometheus.Labels{"handler": "push"},
+		},
+		[]string{"code"},
```

Comment:
Now that we warned so much about partitioning histograms, how about partitioning by method here? That is more likely to make sense, and has a more limited cardinality.